include/envoy/ssl/private_key/private_key_config.h

```diff
@@ -12,10 +12,29 @@ namespace Ssl {
 class PrivateKeyMethodProviderInstanceFactory {
 public:
   virtual ~PrivateKeyMethodProviderInstanceFactory() = default;
+
+  /**
+   * Create a particular PrivateKeyMethodProvider implementation. If the implementation is
+   * unable to produce a PrivateKeyMethodProvider with the provided parameters, it should throw
+   * an EnvoyException. The returned pointer should always be valid.
+   * @param config supplies the custom proto configuration for the PrivateKeyMethodProvider
+   * @param context supplies the factory context
+   */
   virtual PrivateKeyMethodProviderSharedPtr createPrivateKeyMethodProviderInstance(
       const envoy::api::v2::auth::PrivateKeyProvider& config,
       Server::Configuration::TransportSocketFactoryContext& factory_context) PURE;
+
+  /**
+   * @return std::string the identifying name for a particular implementation of
+   * PrivateKeyMethodProvider produced by the factory.
+   */
   virtual std::string name() const PURE;
+
+  /**
+   * @return std::string the identifying category name for objects created by this factory.
+   * Used for automatic registration with FactoryCategoryRegistry.
+   */
+  static std::string category() { return "private_key_providers"; };
```

Comment:
It looks like there was a test previously that covered a test provider (`TestPrivateKeyMethodFactory`). @jpeach @ipuustin  how come that test didn't fail without the category defined? I'm wondering if it has something to do with the factory injection code used in the test? (see https://github.com/envoyproxy/envoy/blob/bb76ead628035142490a852ccace2cdbd50fe0db/test/extensions/transport_sockets/tls/ssl_socket_test.cc#L259)

Reply:
The test doesn't use the factory registry, it [directly declares an instance](https://github.com/envoyproxy/envoy/blob/master/test/extensions/transport_sockets/tls/ssl_socket_test.cc#L258) of the test factory, so the category is not required there. 

`InjectFactory` only requires the injectee to implement `name()` and doesn't enforce any requirements on the base factory type.

Reply:
Can we possible have the test injection utility class use category somehow so it will fail to compile? It would be nice to avoid cases like this in the future? WDYT?

Reply:
Maybe something like this:
```
diff --git test/test_common/registry.h test/test_common/registry.h
index 0cdcd8f5a..cf8d1bc31 100644
--- test/test_common/registry.h
+++ test/test_common/registry.h
@@ -15,6 +15,7 @@ namespace Registry {
 template <class Base> class InjectFactory {
 public:
   InjectFactory(Base& instance) : instance_(instance) {
+    EXPECT_TRUE(FactoryCategoryRegistry::isRegistered(Base::category()));
     displaced_ = Registry::FactoryRegistry<Base>::replaceFactoryForTest(instance_);
   }

```

That would break any injection for factories that don't register categories though.

Reply:
Yeah I think that's fine. Let's just force breakage and defining category IMO. Thank you!

/wait

Reply:
Ok, I'll add the fix by @jpeach and also register the `TestPrivateKeyMethodFactory` properly.

Reply:
It turned out that there are other injected factories which are not registered, and I think the idea in the factory injection is that the factories do not need to be registered? I changed `InjectFactory` check so that it only checks that the category is non-empty (and at the same time it fails to compile if the `category()` function doesn't exist).